### ISSUE TITLE
Addition of OHEM sampler

### ISSUE THREAD

**Issue by rush2406** [pytorch/vision#4286]
## 🚀 Feature
Addition of OHEM (online hard example mining) sampler

## Motivation

Currently, a random sampler is being used in RoIHeads. However, other samplers have been shown to perform better than random sampling and heuristics. Thus, this motivates the need to introduce other samplers for example OHEM.

## Pitch

There can be flag/option introduced to use a particular sampler. Based on the flag, the corresponding sampler would be used.

**Comment by datumbox** [comment]
I agree it would be useful. The key challenge is providing a good generic API what can be used across multiple use-cases without sacrificing speed.

I faced something similar when I was writing the implementation of the SSD model which includes a Hard negative sampling component on the loss. My original reaction was to create a sampler similar to [BalancedPositiveNegativeSampler](https://github.com/pytorch/vision/blob/master/torchvision/models/detection/_utils.py#L11) nevertheless it became clear that writing a [custom implementation](https://github.com/pytorch/vision/blob/3e7653c8c4b5eb0c854251a0fb86be18de55bc68/torchvision/models/detection/ssd.py#L268-L276) in the SSD loss that exploits precomputed stats was much more efficient. I'm sure that with some thought one could provide a good enough generic implementation but it's worth discussing first the potential approaches to cover the different use-cases.

**Comment by rush2406** [comment]
@datumbox probably I could create a pull request for OHEM. Based on that, the changes which are required for including new samplers would become clearer?

**Comment by datumbox** [comment]
@rush2406 If you would like to start an experimental PR to investigate options that's totally fine. It's a great way to speed up conversations and might even be merged if it has no breaking changes. I do suspect that it will be BC-breaking though and hence why I marked this proposal with "discussion needed".

**Comment by fmassa** [comment]
Hi @rush2406 

Thanks for the proposal!

Correct me if I'm wrong, but have OHEM been used in newer models since its introduction, or have something like `focal_loss` replaced most of its usages nowadays?

**Comment by rush2406** [comment]
Hi @fmassa 
As far I as know, focal loss is usually used in the 1 stage detectors to counter imbalance between foreground and background proposals which is not an issue in 2 stage detectors. OHEM would be more useful for 2 stage detectors like faster rcnn.

**Comment by rush2406** [comment]
@datumbox @fmassa could you please review #4305 and provide feedback?

**Comment by fmassa** [comment]
Hi @rush2406 

From my understanding one can exchange OHEM with Focal Loss. Indeed, in the [RetinaNet paper Table 1](https://arxiv.org/pdf/1708.02002.pdf) they compare both approaches and shows that Focal Loss improves on top of OHEM by over 3 points
![image](https://user-images.githubusercontent.com/9110200/130819346-5ca1979a-0aa2-49bf-86cf-cbc65242507c.png)

Have you seen OHEM being used in recent models?

**Comment by rush2406** [comment]
Hi @fmassa 
OHEM and focal loss do pretty much the same thing except that focal loss retains all proposals unlike OHEM. It's also true that focal loss has become much more popular than OHEM. However, my objective here was to draw attention towards the possibility of introducing various samplers, one of which is OHEM. There are others as well like IoU based sampling, prime sampling etc. Hence, I created a sample PR to provide sort of an idea of the changes required with the current codebase.

**Comment by datumbox** [comment]
> From my understanding one can exchange OHEM with Focal Loss. 

OHEM is useful in problems where the negative examples are too many to consider all. Admittedly there are more such examples in NLP and Recommender Systems than in Computer Vision.

> my objective here was to draw attention towards the possibility of introducing various samplers

I think this was valuable as it exposes a limitation of the API, which is that the Sampler is being called in a place where important information is not yet available (the loss in this case). Though we might decide it's not worth redesigning the API, it's good to have a clear example that showcases the limitation of our current approach.
